lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -763,52 +906,36 @@ bool HandshakeBuilder::visitHandshake(ControlMergeOp op) {
   // index to the win wire. If the won register is set, just use that. In
   // the case that won is not set and no input is valid, set a sentinel value to
   // indicate no winner was chosen. The constant values are remembered in a map
   // so they can be re-used later to assign the arg ready outputs.
-  DenseMap<unsigned, Value> argIndexValues;
-  auto priorityMux = noWinner;
-  for (unsigned i = argValid.size(); i > 0; --i) {
-    unsigned inputIndex = i - 1;
-    auto constIndex = createConstantOp(indexType, APInt(indexBits, inputIndex),
-                                       insertLoc, rewriter);
-    priorityMux = rewriter.create<MuxPrimOp>(
-        insertLoc, indexType, argValid[inputIndex], constIndex, priorityMux);
-    argIndexValues[inputIndex] = constIndex;
-  }
+  DenseMap<size_t, Value> argIndexValues;
+  auto priorityArb = createPriorityArbiter(argValid, noWinner, argIndexValues,
+                                           insertLoc, rewriter);
 
-  priorityMux = rewriter.create<MuxPrimOp>(
-      insertLoc, indexType, hadWinnerCondition, won, priorityMux);
-  rewriter.create<ConnectOp>(insertLoc, win, priorityMux);
+  priorityArb = rewriter.create<MuxPrimOp>(
+      insertLoc, indexType, hadWinnerCondition, won, priorityArb);
+
+  rewriter.create<ConnectOp>(insertLoc, win, priorityArb);
 
   // Create the logic to assign the result and control outputs. The result valid
   // output will always be assigned, and if isControl is not set, the result
   // data output will also be assigned. The control valid and data outputs will
-  // always be assigned. The win wire from the arbiter is reinterpreted as
-  // unsigned. The unsigned wire is used to index into a tree of muxes to select
-  // the chosen input's signal(s), and is fed directly to the control output.
-  // Both the result and control valid outputs are gated on the win wire being
-  // set to something other than the sentinel value.
-  auto winUnsignedType = UIntType::get(context, indexBits - 1);
-
-  Value winUnsigned =
-      rewriter.create<BitsPrimOp>(insertLoc, win, indexBits - 2, 0);
-  winUnsigned =
-      rewriter.create<AsUIntPrimOp>(insertLoc, winUnsignedType, winUnsigned);
-
+  // always be assigned. The win wire from the arbiter is used to index into a
+  // tree of muxes to select the chosen input's signal(s), and is fed directly
+  // to the control output. Both the result and control valid outputs are gated
+  // on the win wire being set to something other than the sentinel value.
   auto resultNotEmitted =
       rewriter.create<NotPrimOp>(insertLoc, bitType, resultEmitted);
 
   auto resultValidWire =
-      createMuxTree(argValid, winUnsigned, insertLoc, rewriter);
+      createOneHotMuxTree(argValid, win, insertLoc, rewriter);
   resultValidWire = rewriter.create<AndPrimOp>(
       insertLoc, bitType, resultValidWire, hasWinnerCondition);
   resultValidWire = rewriter.create<AndPrimOp>(
```

Comment:
It seems the case of ControlMergeOp is similar to MergeOp, `resultValidWire` can be driven by `hasWinnerCondition & ResultNotEmitted` directly. Although here `win` is possible to be driven by `won`, but I don't think it should happen that the `valid` signal of the winner argument is down. Will this make any sense?

Reply:
Yep, makes sense. That is already how the control valid is driven, and we can use the same idea for the result valid.